src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/logic/WorkflowLogic.kt

```diff
@@ -34,18 +46,79 @@ class OrderlyWebWorkflowLogic : WorkflowLogic
         val columnCount = headers.count()
         val paramNames = headers.drop(1)
 
-        return rows.drop(1).mapIndexed { rowIdx, row ->
+        val errors: MutableList<String> = mutableListOf()
+        val reports = rows.drop(1).mapIndexed { rowIdx, row ->
             if (row.count() != columnCount)
             {
-                throw BadRequest(
-                        "Report row ${rowIdx + 1} should contain $columnCount values, ${row.count()} values found")
+                errors.add("Report row ${rowIdx + 1} should contain $columnCount values, ${row.count()} values found")
             }
 
             val reportName = row[0]
             val parameters = paramNames.mapIndexed { i, name ->
-                name to row[i + 1]
-            }.filter { it.second.isNotBlank() }.toMap()
+                name to if (row.count() > i + 1) row[i + 1] else ""
+            }.filter { it.second.isNotEmpty() }.toMap()
+
             WorkflowReportWithParams(reportName, parameters)
         }
+
+        val errorTemplate = { index: Int, msg: String -> "Report row $index: $msg" }
+        errors += validateWorkflowReports(reports, branch, commit, orderly, errorTemplate)
+
+        if (errors.isNotEmpty())
+        {
+            throw BadRequest(errors)
+        }
+
+        return reports
+    }
+
+    private fun validateWorkflowReports(
+        reports: List<WorkflowReportWithParams>,
+        branch: String?,
+        commit: String?,
+        orderly: OrderlyServerAPI,
+        errorTemplate: (index: Int, msg: String) -> String
```

Comment:
Using this errorTemplate just in case this logic might end up being repurposed for non-csv validation, in which case 'Report row..' prefix wouldn't be appropriate.

Reply:
Ah, ignore my comment about this being a class level val then.